src_template/ffurls.1
.\"
.\" This manpage is a part of __PROGRAM_NAME__ __PROGRAM_VERSION__
.\"
.\" __PROGRAM_COPYRIGHT__ __PROGRAM_AUTHOR__ __PROGRAM_AUTHOR_EMAIL__
.\"
.\" This program is free software: you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation, either version 3 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program.  If not, see <http://www.gnu.org/licenses/>.
.\"

.TH __PROGRAM_NAME__ "1" "__PROGRAM_DATE__" "__PROGRAM_NAME__ __PROGRAM_VERSION__" "User Commands"

.SH NAME

__PROGRAM_NAME__ \- save tabs opened in Firefox to a file in some format.

.SH SYNOPSIS

.B __FNAME_GENERAL_SH__
[ \fB\-\-text\fR | \fB\-\-org\fR | \fB\-\-html\fR ]

.SH DESCRIPTION

.PP
Save open browser Firefox tabs (title and url) to the file with a given format.

It takes Firefox file with actual configuration of opened tabs and
windows, finds titles and urls there and saves these pairs to a file
in a selected format.

Some file formats are mutable, some not and some are both mutable and
immutable.

Available formats:

  text -- a text file
          a raw format, it is handy for automatic processing

  html -- a html page
          it is handy for read, but it is closed for editions

   org -- org-file for Emacs
          it is handy for read, also it is handy for editions
          (recommended)

.PP
The default format from config file is used.

.SH OPTIONS

.TP
\fB\-\-text\fR
save output to text format

.TP
\fB\-\-org\fR
save output to org format

.TP
\fB\-\-html\fR
save output to html format

.TP
\fB\-\-version\fR
print the program version

.SH AUTHOR

Written by __PROGRAM_AUTHOR__

.SH "REPORTING BUGS"

Report bugs and feature requests to
.br
__PROGRAM_WEBSITE_SECOND__

.SH COPYRIGHT

__PROGRAM_COPYRIGHT__ __PROGRAM_AUTHOR__ __PROGRAM_AUTHOR_EMAIL__
.br
__PROGRAM_LICENSE__

.SH FILES

.TP
\fI __DNAME_ETC_CONFIG__/__FNAME_ETC_CONFIG__ \fR
configuration for all users

.TP
\fI __DNAME_HOME_CONFIG__/__FNAME_HOME_CONFIG__ \fR
configuration for the current user

.SH "SEE ALSO"

.BR __PROGRAM_NAME__ (5)


NEWS

This is the change log of __PROGRAM_NAME__ __PROGRAM_VERSION__

Change Log
All notable changes to this project will be documented in this file.

Unreleased

Added

Changed

Fixed

[1.1.2] - 2018-10-30

Changed
- Program files moved from the user local directory to system
  directories.
- Remove suffixes .sh and .py from program files names.

[1.1.1] - 2018-09-22

Added
- Now, the user can create his own config file in his home directory,
  overriding setting of the general config file for all users.

[1.1.0] - 2018-09-19

Added
- Parsing for Firefox 60
  For this browser version at first it decompresses file before parsing.

Changed
- Four shell-scripts squashed to one shell-script with parameters.

[1.0.3hf1] - 2017-03-23

Changed
- Removed short shell-script names while installation
- Removed installation to $HOME directory
- Program name can be changed in all project files

[1.0.3] - 2017-02-16

Added
- Secondary web-site in README

[1.0.3rc1] - 2017-02-11

Added
- Requirements information in README
- Troubleshooting information in README
- Real author's address and website
- Project description in README.md

Fixed
- Incorrect work when no input file
- Incorrect work when cannot write output file

[1.0.2] - 2017-01-19

Added
- Change log in file NEWS

[1.0.2rc4] - 2017-01-16

Added
- README
- Add docs building by templates and installation
- File configure to create right Makefile automatically

Changed
- Put Makefile under the license
- Makefile now is generated by m4 template

Fixed
- The bug with building project with root permissions
- The bug with the static home directory in Makefile

[1.0.2rc3] - 2017-01-11

Added
- The license file

Changed
- Put scripts under the license

[1.0.2rc2] - 2017-01-10

Fixed
- The bug of the cleared browser cache
- The bug of an absent title

[1.0.2rc1] - 2017-01-10

Added
- Program m4 for scripts generation

Changed
- In org output changed urls to checkbox list items

[1.0.1] - 2016-11-27

Added
- Description and copyright to Makefile

[1.0.1rc4] - 2016-11-27

Added
- Description and copyright to shell-scripts

[1.0.1rc3] - 2016-11-27

Added
- Functions' docstrings

Changed
- Module docscring translated to English

[1.0.1rc2a] - 2016-11-27

Added
- __PROGRAM_NAME__.sh running __PROGRAM_NAME__-org.sh as default
- Project's Makefile
- Version, author and date of the program to __PROGRAM_NAME__.py

[1.0.1rc2] - 2016-11-25

Added
- Option of org output to __PROGRAM_NAME__.py
- __PROGRAM_NAME__-org.sh running __PROGRAM_NAME__.py with html output.

[1.0.1rc1] - 2016-11-24

Added
- Option of html output to __PROGRAM_NAME__.py
- __PROGRAM_NAME__-html.sh running __PROGRAM_NAME__.py with html output. Also it
  increments the file name if the file exists already.

Changed
- __PROGRAM_NAME__.sh renamed to __PROGRAM_NAME__-text.sh
- Previous __PROGRAM_NAME__.sh didn't increment file when repeating the saving.
  Now, it doesn't rewrite the file and make a new file with an
  incremented number in the file name.

[1.0.0] - 2016-05-15

Added
- __PROGRAM_NAME__.py with text output
- __PROGRAM_NAME__.sh running __PROGRAM_NAME__.py with default paths


src_template/ffurls.py
#!/usr/bin/env python3

# __PROGRAM_NAME__ __PROGRAM_VERSION__
#
# __PROGRAM_COPYRIGHT__ __PROGRAM_AUTHOR__ __PROGRAM_AUTHOR_EMAIL__
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

"""Saves tabs opened in Firefox to a file in some format.

It takes Firefox file with actual configuration of opened tabs and
windows, finds titles and urls there and saves these pairs to a file
in a selected format.

Some file formats are mutable, some not and some are both mutable and
immutable.

Available formats:

  text -- a text file
          a raw format, it is handy for automatic processing

  html -- a html page
          it is handy for read, but it is closed for editions

   org -- org-file for Emacs
          it is handy for read, also it is handy for editions
          (recommended)

"""

__version__ = '__PROGRAM_VERSION_NO_V__'
__date__ = '__PROGRAM_DATE__'
__author__ = '__PROGRAM_AUTHOR__ __PROGRAM_AUTHOR_EMAIL__'
__license__ = 'GNU GPLv3'


import sys
import argparse
import json

class SrcNotFoundError(FileNotFoundError):
    """Source file not found exception."""

class DstNotFoundError(FileNotFoundError):
    """Destination file not found exception."""

def get_json_data(fname):
    """Load json-data from file on file system."""
    with open(fname, encoding='utf-8') as fin:
        return json.load(fin)

def get_ff_title_url_pairs(data):
    """Select title and url pairs from Firefox json data."""
    for win in data['windows']:
        for tab in win['tabs']:
            for entry in tab['entries']:
                entry_title = entry.get('title')
                entry_url = entry.get('url')
                yield entry_title, entry_url

def strings_to_file(fname, seq):
    """Save strings from sequence to the file on file system."""
    with open(fname, 'w', encoding='utf-8') as fout:
        for i in seq:
            print(i, file=fout)

def text_to_file(fname, text):
    """Save text to the file on file system."""
    with open(fname, 'w', encoding='utf-8') as fout:
        fout.write(text)

def replace_nones(seq, rep_left, rep_right):
    """Replace in sequence of pairs none values to replacement values.

    For example:

      [(1, 1),
       (None, 1),
       (1, None),
       (None, None)]

      becomes:

      [(1, 1),
       (rep_left, 1),
       (1, rep_right),
       (rep_left, rep_right)]

    """
    for in_left, in_right in seq:
        out_left = in_left if in_left is not None else rep_left
        out_right = in_right if in_right is not None else rep_right
        yield out_left, out_right

def convert_ff_to_txt(ifname, ofname):
    """Convert Firefox tabs to the text file with title and urls."""
    try:
        ffurls = get_ff_title_url_pairs(get_json_data(ifname))
    except FileNotFoundError as exc:
        raise SrcNotFoundError(exc)
    ffurls_replaced = replace_nones(ffurls, 'Unknown title', 'unknown')
    tustrs = ('{}\n{}'.format(t, u) for t, u in ffurls_replaced)
    try:
        strings_to_file(ofname, tustrs)
    except FileNotFoundError as exc:
        raise DstNotFoundError(exc)

def convert_ff_to_html(ifname, ofname):
    """Convert Firefox tabs to the html file with title and urls."""
    try:
        ffurls = get_ff_title_url_pairs(get_json_data(ifname))
    except FileNotFoundError as exc:
        raise SrcNotFoundError(exc)
    ffurls_replaced = replace_nones(ffurls, 'Unknown title', 'unknown')
    html_page = make_html_page(ffurls_replaced)
    try:
        text_to_file(ofname, html_page)
    except FileNotFoundError as exc:
        raise DstNotFoundError(exc)

def convert_ff_to_org(ifname, ofname):
    """Convert Firefox tabs to the org file with title and urls."""
    try:
        ffurls = get_ff_title_url_pairs(get_json_data(ifname))
    except FileNotFoundError as exc:
        raise SrcNotFoundError(exc)
    ffurls_replaced = replace_nones(ffurls, 'Unknown title', 'unknown')
    org_text = make_org_text(ffurls_replaced)
    try:
        text_to_file(ofname, org_text)
    except FileNotFoundError as exc:
        raise DstNotFoundError(exc)

def make_html_page(seq):
    """Make html-page from sequence of (title, url) pairs."""
    fmt = \
"""\
<html>
<head>
  <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
  <title>Saved Firefox Urls</title>
  <style type="text/css">
    body {{ font-size: small; }}
  </style>
</head>
<body>
  <ul>
{}
  </ul>
</body>
</html>
"""
    ul_items = ('    <li><a href="{1}">{0}</a>'.format(t, u)
                for t, u in seq)
    out = fmt.format('\n'.join(ul_items))
    return out

def make_org_text(seq):
    """Make org-text from sequence of (title, url) pairs."""
    fmt = \
"""\
#+STARTUP: showall

* Saved Firefox urls
{}
"""
    items = ('  - [ ] [[{1}][{0}]]'.format(t, u)
             for t, u in seq)
    out = fmt.format('\n'.join(items))
    return out

def get_prog_args():
    """Parse command line arguments to a handy object with attributes."""
    desc = """
    Converts Firefox session json file to
    a text, html or emacs org file with titles and urls.
    """
    parser = argparse.ArgumentParser(description=desc)
    parser.add_argument('--version', '-V',
                        action='version',
                        version='%(prog)s ' + 'v' + __version__)
    parser.add_argument('--license',
                        action='version',
                        version='License: ' + __license__ +
                        ', see more details in file LICENSE'
                        ' or at <http://www.gnu.org/licenses/>.',
                        help='show program\'s license and exit')
    parser.add_argument('-t',
                        choices=['text', 'html', 'org'],
                        default='text',
                        help='type of the output file (default: %(default)s)')
    parser.add_argument('ifname',
                        help='input Firefox session file (recovery.js)')
    parser.add_argument('ofname', help='output urls text, html or org file')
    return parser.parse_args()

def print_error(message):
    """Print an error message to stderr."""
    print('error:', message, file=sys.stderr)

def main():
    """Convert Firefox tabs to text, html or org file."""
    args = get_prog_args()
    ifname = args.ifname
    ofname = args.ofname
    oftype = args.t
    try:
        if oftype == 'text':
            convert_ff_to_txt(ifname, ofname)
        elif oftype == 'html':
            convert_ff_to_html(ifname, ofname)
        elif oftype == 'org':
            convert_ff_to_org(ifname, ofname)
    except SrcNotFoundError:
        print_error('source data file not found: {}'.format(ifname))
        return 1
    except DstNotFoundError:
        print_error('can\'t create output file: {}'.format(ofname))
        return 1
    return 0

if __name__ == '__main__':
    sys.exit(main())
